core/services/orm/tree_traversal/ModelObjNode.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+namespace EventEspresso\core\services\orm\tree_traversal;
+
+use EE_Base_Class;
+use EE_HABTM_Relation;
+use EE_Has_Many_Relation;
+use EventEspresso\core\exceptions\InvalidDataTypeException;
+use EventEspresso\core\exceptions\InvalidInterfaceException;
+use InvalidArgumentException;
+use ReflectionException;
+
+/**
+ * Class ModelObjNode
+ * Wraps a model object and stores which of its model's relations have already been traversed and which haven't.
+ *
+ * @package     Event Espresso
+ * @author         Mike Nelson
+ * @since         $VID:$
+ *
+ */
+class ModelObjNode extends BaseNode
+{
+    /**
+     * @var EE_Base_Class
+     */
+    protected $model_obj;
+
+    /**
+     * @var RelationNode[]
+     */
+    protected $relation_nodes;
+
+    public function __construct($instance)
+    {
+        $this->model_obj = $instance;
+    }
+
+    /**
+     * Creates a relation node for each relation of this model's relations.
+     * Does NOT call `discover` on them yet though.
+     * @since $VID:$
+     * @throws \EE_Error
+     * @throws InvalidDataTypeException
+     * @throws InvalidInterfaceException
+     * @throws InvalidArgumentException
+     * @throws ReflectionException
+     */
+    protected function discover()
+    {
+        $this->relation_nodes = [];
+        foreach ($this->model_obj->get_model()->relation_settings() as $relationName => $relation) {
+            if ($relation instanceof EE_Has_Many_Relation) {
+                $this->relation_nodes[ $relationName ] = new RelationNode($this->model_obj, $relation->get_other_model());
+            } elseif ($relation instanceof EE_HABTM_Relation) {
+                $this->relation_nodes[ $relation->get_join_model()->get_this_model_name() ] = new RelationNode($this->model_obj, $relation->get_join_model());
+            }
+        }
+        ksort($this->relation_nodes);
+    }
+
+
+    /**
+     * Whether this item has already been initialized
+     */
+    protected function isDiscovered()
+    {
+        return $this->relation_nodes !== null && is_array($this->relation_nodes);
```

Comment:
Kk that sounds like a nice-to-have that I'm not opposed to. I've really only done a handful of unit tests, so that's not the end of the world (thanks for giving me the option though). I'll add a bullet point for it so I can do it if time permits. But I like performance.